crates/iota-grpc-api/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+        grpc_checkpoint_summary_tx: tokio::sync::broadcast::Sender<
+            Arc<GrpcCertifiedCheckpointSummary>,
+        >,
+        grpc_checkpoint_data_tx: tokio::sync::broadcast::Sender<Arc<GrpcCheckpointData>>,
+    ) -> Self {
+        Self {
+            state_reader,
+            grpc_checkpoint_summary_tx,
+            grpc_checkpoint_data_tx,
+        }
+    }
+}
+
+// Checkpoint stream item.
+// Note, node::Checkpoint may contain either checkpoint data or summary.
+type CheckpointStreamResult = Result<node::Checkpoint, Status>;
+
+// Helper trait for getting checkpoint data and summaries,
+// intended as an abstraction for Arc<dyn RestStateReader>.
+trait CheckpointReader<T>
+where
+    T: Send + Sync + 'static + Serialize,
+    Self: Send + Sync + 'static,
+{
+    fn get_sequence_number(&self, item: &Arc<T>) -> u64;
+    fn get_item(&self, ix: u64) -> Option<Arc<T>>;
+    fn get_latest(&self) -> Option<u64>;
+
+    fn create_checkpoint_response(&self, item: &Arc<T>, is_full: bool) -> CheckpointStreamResult {
+        BcsData::serialize_from(item)
+            .map(|data| node::Checkpoint {
+                sequence_number: self.get_sequence_number(item),
+                bcs_data: Some(data),
+                is_full,
+            })
+            .map_err(|e| Status::internal(format!("BCS serialization error: {e}")))
+    }
+}
+
+#[derive(Clone)]
+struct Reader {
+    state_reader: Arc<dyn RestStateReader>,
+}
+
+fn get_full_checkpoint_data(
+    state_reader: &std::sync::Arc<dyn RestStateReader>,
+    seq: u64,
+) -> Option<CheckpointData> {
+    let summary = state_reader.get_checkpoint_by_sequence_number(seq)?;
+    let contents = state_reader.get_checkpoint_contents_by_sequence_number(seq)?;
+    Some(state_reader.get_checkpoint_data(summary, contents))
+}
+
+impl CheckpointReader<GrpcCheckpointData> for Reader {
+    fn get_sequence_number(&self, item: &Arc<GrpcCheckpointData>) -> u64 {
+        item.sequence_number()
+    }
+    fn get_item(&self, ix: u64) -> Option<Arc<GrpcCheckpointData>> {
+        get_full_checkpoint_data(&self.state_reader, ix)
+            .map(GrpcCheckpointData::from)
+            .map(Arc::new)
+    }
+    fn get_latest(&self) -> Option<u64> {
+        Some(*self.state_reader.get_latest_checkpoint().sequence_number())
+    }
+}
+
+impl CheckpointReader<GrpcCertifiedCheckpointSummary> for Reader {
+    fn get_sequence_number(&self, item: &Arc<GrpcCertifiedCheckpointSummary>) -> u64 {
+        item.sequence_number()
+    }
+    fn get_item(&self, ix: u64) -> Option<Arc<GrpcCertifiedCheckpointSummary>> {
+        self.state_reader
+            .get_checkpoint_by_sequence_number(ix)
+            .map(|v| GrpcCertifiedCheckpointSummary::from(CertifiedCheckpointSummary::from(v)))
+            .map(Arc::new)
+    }
+    fn get_latest(&self) -> Option<u64> {
+        Some(*self.state_reader.get_latest_checkpoint().sequence_number())
+    }
+}
+
+fn create_checkpoint_stream<T, F>(
```

Comment:
Why not move this into the Reader impl?

Reply:
Thanks for pointing this out. I moved it at [59dcb92](https://github.com/iotaledger/iota/pull/7511/commits/59dcb926c4393d9255d5c57dd343faac022a66de).